signalr-go-server/pkg/signalr/server.go

```diff
@@ -0,0 +1,208 @@
+package signalr
+
+import (
+	"fmt"
+	"reflect"
+	"runtime/debug"
+	"strings"
+)
+
+type Server struct {
+	hubPrototype HubInterface
+	lifetimeManager HubLifetimeManager
+	defaultHubClients defaultHubClients
+	groupManager GroupManager
+}
+
+func NewServer(hubPrototype HubInterface) *Server {
+	lifetimeManager := defaultHubLifetimeManager{}
+	return &Server{
+		hubPrototype:    hubPrototype,
+		lifetimeManager: &lifetimeManager,
+		defaultHubClients: defaultHubClients{
+			lifetimeManager: &lifetimeManager,
+			allCache:        allClientProxy{lifetimeManager: &lifetimeManager},
+		},
+		groupManager: &defaultGroupManager{
+			lifetimeManager: &lifetimeManager,
+		},
+	}
+}
+
+func (s *Server) MessageLoop(conn HubConnection, connectionID string, protocol HubProtocol) {
```

Comment:
Why is this public?

Reply:
I tried to extract all websocket independant code and expose it to external packages to make it possible to build a server for another transport in a simple way. That worked fine with the pipe-based transport in the signalr_test package. Bonus: External package developers would be able to write their own servers with other transports/connections.
But reconsidering it, "other transports" need to be supported by clients. Afaik, these are long polling and SSE (you might know better) and these should be better implemented inside the package. So exposing MessageLoop, HubConnection, HubProtocol really makes no sense

Reply:
I want this too but not in this way. I'd model it similar to the C# implementation The only thing coupled to the transport needs to be the websocket impl. We just need to abstract what a connection looks like and everything else can remain as it was.

Reply:
See [bc57af5](https://github.com/davidfowl/signalr-ports/pull/6/commits/bc57af5e966c7c76eda9de8e96343dc1452c9260)